### PR TITLE
Implement partial literal propagation pass (support 'isinstance')

### PR DESCRIPTION
**Pull request by guilhermeleobas**
As title. With this, Numba can optimize away `isinstance` calls. 

```python
def test_isinstance(a):
    if isinstance(a, (int, float)):
        if isinstance(a, int):
            return a + 1
        if isinstance(a, float):
            return a + 2.0
    elif isinstance(a, str):
        return a + ', world!'
    return 1+2j
```

I am calling it a partial constant propagation because it is simpler than the one used by LLVM, for instance. It uses information from the partial type inference pass (stored in state.typemap) to replace variables whose value can be determined at compile time (i.e. literals).

Based on #7204

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
I guess the CI failure is because azure is running (almost) all tests?

**Comment by gmarkall** [comment]
> I guess the CI failure is because azure is running (almost) all tests?

It seems to be because tests are actually failing - can you clarify what you mean?

**Comment by guilhermeleobas** [comment]
> > I guess the CI failure is because azure is running (almost) all tests?
> 
> It seems to be because tests are actually failing - can you clarify what you mean?

Hey @gmarkall, I thought the errors were because Numba was running the entire test-suite and CI was timing out.

**Comment by guilhermeleobas** [comment]
PR is ready for review

File: numba/tests/test_lists.py [hunk 1]
```diff
@@ -1758,21 +1758,6 @@ def foo(x):
...
-                l = ['a', 1, 2j]
-            else:
-                l = ['b', 2]
-            return l[0]
-
-        with self.assertRaises(errors.TypingError) as raises:
-            foo(100)
-        expect = "Cannot unify LiteralList"
-        self.assertIn(expect, str(raises.exception))
-
```
Comment (stuartarchibald) [inline numba/tests/test_lists.py:1775]: This test should still work. I think this patch has had the side effect of specialising on the literal value of inputs:
```python
from numba import njit
import numpy as np

@njit
def foo(x):
    if x + 1 > 3:
        l = ['a', 1, 2j]
    else:
        l = ['b', 2]
    return l[0]

print(foo(100))  # 'a'
print(foo(-100)) # also 'a' !
```
note that `foo` will have the signature `(int64,)` but the IR will have over-specialised on the value `100`.

File: numba/tests/test_ir_inlining.py [hunk 2]
```diff
@@ -37,17 +44,41 @@ class InlineTestPipeline(CompilerBase):
...
+
+        typed_passes = dpb.define_typed_pipeline(state)
+        pm.passes.extend(typed_passes.passes)
+
+        lowering_passes = dpb.define_nopython_lowering_pipeline(state)
+        pm.passes.extend(lowering_passes.passes)
+
+        pm.add_pass_after(PreserveIR, IRLegalization)
+
+        pm.finalize()
```
Comment (stuartarchibald) [inline numba/tests/test_ir_inlining.py:80]: What's the purpose of this change?

File: numba/tests/test_builtins.py [hunk 3]
```diff
@@ -948,6 +959,23 @@ def test_sum_exceptions(self):
...
+            5.0,      # float
+            "Hello",  # string
+            1j,       # complex
+        )
+
+        for inpt in inputs:
+            expected = pyfunc(inpt)
+            got = cfunc(inpt)
+            self.assertEqual(expected, got)
+
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py:977]: Would be good to add a load more tests here, for example:
```python
from numba import njit
import numpy as np

@njit
def foo(x):
    if isinstance(x, (int, int)):
        print('is int')
    else:
        print('is not int')

print(foo(100))
```
this currently prints "is not int", which is a bug?!

This is entirely illegal and yet still compiles and runs:
```python
from numba import njit
import numpy as np

@njit
def foo(x):
    if isinstance(x, ('foo',)):
        print('is int')
    else:
        print('is not int')

print(foo(100))
```
and maybe not for this PR, but it'd be good to consider how to get things like:
```python
from numba import njit
import numpy as np

@njit
def foo(x):
    if isinstance(x, tuple):
        print('A')
    else:
        print('B')

foo(100)
```
to work.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this is a great start. I've left a few comments on the PR for things to start debugging/taking a look at. Thanks again!

File: numba/tests/test_ir_inlining.py (hunk 2 above)
Comment (guilhermeleobas) [reply numba/tests/test_ir_inlining.py:80]: I am changing the default nopython pipeline to include the partial constant propagation pass. Which optimizes the IR and breaks existing tests that expect the IR to have certain instructions/certain shape (i.e., `test_inlining_models`).

File: numba/tests/test_lists.py (hunk 1 above)
Comment (guilhermeleobas) [reply numba/tests/test_lists.py]: The problem with this test case is that `typemap` holds information that `l[0] == 'a'`. I've put a check that ignores `static_getitem` but not sure if this is the best approach

File: numba/tests/test_builtins.py (hunk 3 above)
Comment (guilhermeleobas) [reply numba/tests/test_builtins.py:984]: I've fixed the first two cases. In the last case, what is `tuple`?

**Comment by guilhermeleobas** [comment]
Thanks for the review, Stuart. I've address your comments.

File: numba/tests/test_builtins.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/tests/test_builtins.py:984]: > I've fixed the first two cases. In the last case, what is `tuple`?

Thanks, feel free to try more variations! `tuple` here is literally the `tuple` builtin function:
```pycon
In [9]: tuple?
Init signature: tuple(iterable=(), /)
Docstring:     
Built-in immutable sequence.
```

Comment (guilhermeleobas) [reply numba/tests/test_builtins.py:984]: Oh, got it. You mean, any container like `list`, `dict`, `set`, `tuple`, ...

**Comment by guilhermeleobas** [comment]
I've added `isinstance` support for containers like `list` and `tuple` and jitclasses.

**Comment by gmarkall** [comment]
@guilhermeleobas Could you please use merge commits and non-force pushes once you've received review comments? The review history tends to get a bit confused by force-pushes.

**Comment by guilhermeleobas** [comment]
This PR requires an updated version of the dead branch prune pass to reconstruct PHI nodes. cc @stuart

**Comment by guilhermeleobas** [comment]
Out of the 5 failures, 4 are related to DWARF info and one is an error related to changes introduced in this PR.

**Comment by stuartarchibald** [comment]
> Out of the 5 failures, 4 are related to DWARF info and one is an error related to changes introduced in this PR.

I suggest dropping the changes to lowering: https://github.com/numba/numba/pull/7435/files#diff-b62eb18df90934b3331e32a080b3dc6ffa83f3e040bd1af798f37eedb0aac6a2R43-R338 (in lowering.py), the patch technically fixes a couple of possible bugs but I don't think it's critical for the needs of this PR. That will hopefully fix the DWARF problems.

**Comment by guilhermeleobas** [comment]
Current failure is from `debuginfo`. Also, the `AU.add_dependency` doesn't seem to be working properly. I suspect this is a bug in the current pass manager implementation, as `LiteralPropagationPass` requires `SSA` to be executed before, but it is not

**Comment by stuartarchibald** [comment]
> Current failure is from `debuginfo`. Also, the `AU.add_dependency` doesn't seem to be working properly. I suspect this is a bug in the current pass manager implementation, as `LiteralPropagationPass` requires `SSA` to be executed before, but it is not

@guilhermeleobas if you have a fix for the analysis usage dep problem, please could you open a separate PR for it with reproducing test etc so as to get a fix in early? Many thanks.

File: numba/cpython/builtins.py [hunk 4]
```diff
@@ -672,3 +673,53 @@ def impl(func, iterable):
...
+            return true_impl
+        elif isinstance(numba_typ, types.Container) and \
+                numba_typ.key[0] == types.Any:
+            # check for containers (list, tuple, dict, ...)
+            if isinstance(var, numba_typ.__class__) or \
+                (isinstance(var, types.BaseTuple) and \
+                    isinstance(numba_typ, types.BaseTuple)):
+                return true_impl
+
+    return false_impl
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py]: @stuartarchibald, I've pushed a fix to the type inference step. But I think in both cases you posted, Numba shouldn't compile the code since it cannot determine at compile-time the type of `z`.

```python
@njit
def foo(x):
    if x > 4:  # runtime information
        z = 10
    else:
        z = 'a'
    ....
```

File: numba/tests/test_lists.py [hunk 5]
```diff
@@ -1770,7 +1770,7 @@ def foo(x):
 
         with self.assertRaises(errors.TypingError) as raises:
             foo(100)
-        expect = "Cannot unify LiteralList"
+        expect = "Cannot infer"
```
Comment (stuartarchibald) [inline numba/tests/test_lists.py:1773]: I'm not sure these changes are quite right? I think this is a unification problem, not an inference problem, the types of the `l` variables can be inferred, however `l` cannot be unified across the branch as there is no type which encompasses the version if true and the version if false.

**Comment by guilhermeleobas** [comment]
Error is not related to this PR!

**Comment by sklam** [comment]
The following test case is having an unexpected result:

```python
from numba import njit

@njit
def first(a, b):
    return a

@njit
def foo(x):
    if isinstance(x, int):
        y = first(0xcafe + x - x, # so that it's not a literal
                  x)
    elif isinstance(x, float):
        y = first(3.14, x)
    # elif isinstance(x, str):
    #     y = first("NAN", x)
    # the rest depends on knowing the type of `y`
    if isinstance(y, int):
        print("INT")
    elif isinstance(y, float):
        print("FLOAT")
    elif isinstance(y, str):
        print("STR")

    return y


@njit
def bar_int(x):
    z = foo(x + 0xcafe)
    return z


@njit
def bar_float(x):
    z = foo(x + 3.14)
    return z

@njit
def bar_str(x):
    z = foo(str(x))
    return z

assert bar_int(2) ==  0xcafe
assert bar_float(2.0) == 3.14
assert bar_str("3") == "NAN"


print(bar_int.signatures)
print(bar_float.signatures)
print(bar_str.signatures)
```

The call to `bar_int(2)` is triggering `print("FLOAT")`.

If you uncomment the `elif isinstance(x, str):` branch, you will get the expected result. 

The issue has to do with the `PhiConstraint` allowing `y` to be typed as `float` when unifying `int` and `float`.  Such behavior is not wrong given the pre-`isinstance` semantic. Now, with the ability to branch on type, i don't know how we can support the expected behavior.

File: numba/tests/test_builtins.py [hunk 6]
```diff
@@ -948,6 +1030,64 @@ def test_sum_exceptions(self):
...
+        pyfunc = isinstance_usecase_numba_types
+        cfunc = jit(nopython=True)(pyfunc)
+
+        inputs = (
+            (types.int32(1), 'int32'),
+            (types.int64(2), 'int64'),
+            (types.float32(3.0), 'float32'),
+            (types.float64(4.0), 'float64'),
+            (types.complex64(5j), 'no match'),
+            (typed.List([1, 2]), 'typed list'),
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py:1065]: `typed.Dict`?

Comment (stuartarchibald) [reply numba/tests/test_builtins.py:1065]: Also, should this sort of thing work?
```python
from numba import njit
import numpy as np

@njit
def foo():
    a = {'a': 10, 'b': np.zeros(3)}
    if isinstance(a, dict):
        print(True)
    else:
        print(False)

print(foo())
```
?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates to this and for doing the huge amount of debugging/research into type inference propagating types through phi nodes. I've given this another review and have tested it some more, comments are inline, it's a load more robust to strange phi-related failure modes now so the research has paid off! I've asked @sklam to take a look at the part of the code I have written. Thanks again for your work on this!

File: numba/core/untyped_passes.py [hunk 7]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+                    if guard(get_definition, state.func_ir, target.name) is None:  # noqa: E501
+                        continue
+
+                if isinstance(value, ir.Expr) and value.op == 'call':
+                    for arg in value.args:
+                        typs = state.func_ir._definitions[arg.name]
+                        is_phi = any([isinstance(typ, ir.Expr) and typ.op == 'phi' for typ in typs])  # noqa: E501
+                        is_isinstance_call = state.func_ir.get_definition(value.func.name).name == 'isinstance'  # noqa: E501
+                        if is_phi and is_isinstance_call:
+                            msg = 'isinstance cannot handle PHI nodes'
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1472]: Is there a place where I can get the name of the python variable from an IR variable? Or can I just use the IR variable name?

File: numba/core/untyped_passes.py [hunk 8]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+                # 2) Don't propagate literal values that are not primitives
+                if isinstance(value, ir.Expr) and \
+                        value.op in ('cast', 'build_map', 'build_list',
+                                     'build_tuple', 'build_set'):
+                    continue
+
+                target = assign.target
+                if not state.flags.enable_ssa:
+                    if guard(get_definition, state.func_ir, target.name) is None:  # noqa: E501
+                        continue
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1464]: Yes, this pass depends on SSA. But there is one case (inlining) where SSA is disabled

File: numba/core/untyped_passes.py [hunk 9]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+
+                if isinstance(value, ir.Expr) and value.op == 'call':
+                    for arg in value.args:
+                        typs = state.func_ir._definitions[arg.name]
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1468]: Yes, `_definitions` is necessary. This snippet of code is checking if any of the args to `isinstance` is a PHI node. Naming the variable as `typs` was probably a bad choice of word.

File: numba/core/untyped_passes.py [hunk 10]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+                        is_phi = any([isinstance(typ, ir.Expr) and typ.op == 'phi' for typ in typs])  # noqa: E501
+                        is_isinstance_call = state.func_ir.get_definition(value.func.name).name == 'isinstance'  # noqa: E501
+                        if is_phi and is_isinstance_call:
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1471]: exactly!

File: numba/cpython/builtins.py [hunk 11]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+        raise NumbaTypeError(
+            f'isinstance cannot handle optional types. '
+            f'Got "{var_ty}"')
+
+    # register container types (List, Dict, ...) with types.Any as base type
+    as_numba_type.register(tuple, types.Tuple((types.undefined,)))
+    as_numba_type.register(list, types.List(types.undefined))
+    # set requires the type to be hashable
+    as_numba_type.register(set, types.Set(types.undefined))
+    as_numba_type.register(types.ListType, types.ListType(types.undefined))
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:699]: I can safely remove this code and do it different, if needed

File: numba/tests/test_builtins.py [hunk 12]
```diff
@@ -220,6 +220,88 @@ def sum_kwarg_usecase(x, start=0):
...
+            return a + 2.0, 'float'
+    elif isinstance(a, str):
+        return a + ", world!", 'str'
+    elif isinstance(a, complex):
+        return a.imag, 'complex'
+    elif isinstance(a, (tuple, list)):
+        if isinstance(a, tuple):
+            return 'tuple'
+        else:
+            return a.reverse(), 'list'
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_builtins.py:237]: Sort of! I'm calling a method that only exists in a list to make sure the branch pruning is working correctly.

File: numba/tests/test_builtins.py [hunk 13]
```diff
@@ -948,6 +1030,64 @@ def test_sum_exceptions(self):
...
+            (typed.List([1, 2]), 'typed list'),
+        )
+
+        for inpt, expected in inputs:
+            got = cfunc(inpt)
+            self.assertEqual(expected, got)
+
+        pyfunc_2 = isinstance_usecase_numba_types_2
+        cfunc_2 = isinstance_usecase_numba_types_2
+        self.assertEqual(pyfunc_2, cfunc_2)
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_builtins.py:1074]: Fixed

File: numba/cpython/builtins.py [hunk 14]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+        return False
+
+    var_ty = as_numba_type(var)
+
+    if isinstance(var_ty, types.Optional):
+        raise NumbaTypeError(
+            f'isinstance cannot handle optional types. '
+            f'Got "{var_ty}"')
+
+    # register container types (List, Dict, ...) with types.Any as base type
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:694]: Yes, it works!

File: numba/core/untyped_passes.py [hunk 15]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+                target = assign.target
+                if not state.flags.enable_ssa:
+                    if guard(get_definition, state.func_ir, target.name) is None:  # noqa: E501
+                        continue
+
+                if isinstance(value, ir.Expr) and value.op == 'call':
+                    for arg in value.args:
+                        typs = state.func_ir._definitions[arg.name]
+                        is_phi = any([isinstance(typ, ir.Expr) and typ.op == 'phi' for typ in typs])  # noqa: E501
+                        is_isinstance_call = state.func_ir.get_definition(value.func.name).name == 'isinstance'  # noqa: E501
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1470]: Fixed!

File: numba/tests/test_builtins.py (hunk 6 above)
Comment (guilhermeleobas) [reply numba/tests/test_builtins.py:1078]: Done!

File: numba/cpython/builtins.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:694]: Done!

File: numba/cpython/builtins.py [hunk 16]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+        elif isinstance(typ, types.ClassType):
+            key = typ  # jitclasses
+        else:
+            key = typ.key
+
+        # corner cases for bytes, range, ...
+        # avoid registering those types on `as_numba_type`
+        types_not_registered = {
+            bytes: types.Bytes,
+            range: types.RangeType,
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:723]: done!

File: numba/cpython/builtins.py [hunk 17]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+        if key in types_not_registered:
+            if (key == bytes and isinstance(var_ty, types.Bytes)) or \
+                (key == range and isinstance(var_ty, types.RangeType)):
+                return true_impl
+            continue
+
+        numba_typ = as_numba_type(key)
+
+        if var_ty == numba_typ:
+            return true_impl
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py]: Done!

File: numba/cpython/builtins.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:723]: range is tested on `isinstance_usecase_numba_types_2`

File: numba/tests/test_debuginfo.py [hunk 18]
```diff
@@ -317,7 +317,7 @@ def test_DILocation_undefined(self):
         """ Tests that DILocation information for undefined vars is associated
         with the line of the function definition (so it ends up in the prologue)
         """
-        @njit(debug=True)
+        @njit(debug=True, no_cfunc_wrapper=True)
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_debuginfo.py:320]: No. I've removed it

File: numba/core/untyped_passes.py [hunk 19]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
...
+                    # replace assign instruction by ir.Const(lit) iff
+                    # lit is a literal value
+                    rhs = ir.Const(lit.literal_value, assign.loc)
+                    new_assign = ir.Assign(rhs, target, assign.loc)
+
+                    # replace instruction
+                    block.insert_after(new_assign, assign)
+                    block.remove(assign)
+
+                    changed = True
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py]: You mean, move this part to a function?

File: numba/tests/test_builtins.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_builtins.py]: I've removed the `a.reverse()` call

File: numba/cpython/builtins.py [hunk 20]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+    as_numba_type.register(types.ListType, types.ListType(types.undefined))
+
+    t_typs = typs
+
+    if isinstance(t_typs, types.UniTuple):
+        # corner case - all types in isinstance are the same
+        t_typs = (t_typs.key[0])
+
+    if not isinstance(t_typs, types.containers.Tuple):
+        t_typs = (t_typs, )
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:708]: This code is not actually valid in Python. `foo.py_func(100)` raises the following error:
```python
Traceback (most recent call last):
  File "a.py", line 16, in <module>
    print(foo.py_func(100))
  File "a.py", line 7, in foo
    if isinstance(x, [int, str]):
TypeError: isinstance() arg 2 must be a type or tuple of types
```

File: numba/cpython/builtins.py [hunk 21]
```diff
@@ -673,3 +675,78 @@ def impl(func, iterable):
...
+
+    var_ty = as_numba_type(var)
+
+    if isinstance(var_ty, types.Optional):
+        raise NumbaTypeError(
+            f'isinstance cannot handle optional types. '
+            f'Got "{var_ty}"')
+
+    # register container types (List, Dict, ...) with types.Any as base type
+    as_numba_type.register(tuple, types.Tuple((types.undefined,)))
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:695]: Is namedtuple supported inside a jitted function?

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cpython/builtins.py:695]: Yes. e.g.
```python
from numba import njit
import collections

NT = collections.namedtuple('NT', 'x')

nt = NT(1)

@njit
def foo(a):
    print("arg", a)
    print("new one", NT('string'))

foo(nt)
```

[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:695]: Done!

File: numba/core/untyped_passes.py [hunk 22]
```diff
@@ -1426,6 +1427,152 @@ def run_pass(self, state):
...
+        from numba.core.typed_passes import PartialTypeInference
+        pm = PassManager("literal_propagation_subpipeline")
+
+        pm.add_pass(PartialTypeInference, "performs partial type inference")
+        pm.add_pass(PropagateLiterals, "performs propagation of literal values")
+
+        # rewrite consts / dead branch pruning
+        pm.add_pass(RewriteSemanticConstants, "rewrite semantic constants")
+        pm.add_pass(DeadBranchPrune, "dead branch pruning")
+        pm.add_pass(DeadCodeElimination, "dead code elimination")
```
Comment (stuartarchibald) [inline numba/core/untyped_passes.py:1566]: What's the reason for adding this? I've a vague memory of it being unsafe for use on IR that isn't fully typed as alias analysis is needed.

File: numba/cpython/builtins.py [hunk 23]
```diff
@@ -673,3 +675,81 @@ def impl(func, iterable):
...
+            continue
+
+        if isinstance(typ, types.TypeRef):
+            # Case for TypeRef (i.e. isinstance(var, typed.List))
+            #      var_ty == ListType[int64] (instance)
+            #         typ == types.ListType  (class)
+            return true_impl if type(var_ty) is key else false_impl
+        elif isinstance(typ, types.NamedTupleClass):
+            # Case for NamedTuples
+            return true_impl if issubclass(var_ty.__class__, types.BaseNamedTuple) else false_impl
```
Comment (stuartarchibald) [inline numba/cpython/builtins.py:737]: I'm not sure that this is quite right because it's just checking in the typing domain? e.g. This breaks:
```python
from numba import njit
import collections

NT = collections.namedtuple('NT', 'x')
ZT = collections.namedtuple('ZT', 'x')

nt = NT(1)

@njit
def foo(a):
    return isinstance(a, NT), isinstance(a, ZT)

assert foo(nt) == foo.py_func(nt)

```

Comment (stuartarchibald) [reply numba/cpython/builtins.py:737]: Maybe something similar to this could help?:
```python
            if issubclass(var_ty.__class__, types.BaseNamedTuple) and\
                issubclass(var_ty.instance_class, typs.instance_class):
                return true_impl
            else:
                return false_impl
```

File: numba/core/untyped_passes.py (hunk 22 above)
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1566]: Would it be ok to add support for namedtuples in a follow up PR?

File: numba/cpython/builtins.py [hunk 24]
```diff
@@ -673,3 +675,81 @@ def impl(func, iterable):
...
+        else:
+            numba_typ = as_numba_type(key)
```
Comment (sklam) [inline numba/cpython/builtins.py:739]: Avoid using `as_numba_type(key)` here. Can just check if it is one of the python containers to avoid the `as_numba_type.register` above.

File: numba/core/analysis.py [hunk 25]
```diff
@@ -402,6 +402,15 @@ def resolve_input_arg_const(input_arg_idx):
...
 
+    phi2lbl = dict()
+    phi2asgn = dict()
+    for lbl, blk in func_ir.blocks.items():
+        for stmt in blk.body:
+            if isinstance(stmt, ir.Assign):
+                if isinstance(stmt.value, ir.Expr) and stmt.value.op == 'phi':
+                    phi2lbl[stmt.value] = lbl
+                    phi2asgn[stmt.value] = stmt
+
```
Comment (sklam) [inline numba/core/analysis.py:413]: nitpicking: any reason for this block of code to be so far away from the use of `phiasgn`?

File: numba/core/analysis.py [hunk 26]
```diff
@@ -492,9 +501,71 @@ def resolve_input_arg_const(input_arg_idx):
                         repl_idx = defns.index(cond)
                         defns[repl_idx] = x.value
 
+    # Check post dominators of dead nodes from in the original CFG for use of
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/analysis.py:504]: All the changes here seem unnecessary now that `isinstance` support is reduced.

File: numba/cpython/builtins.py (hunk 24 above)
Comment (guilhermeleobas) [reply numba/cpython/builtins.py:732]: done!

File: numba/core/untyped_passes.py [hunk 27]
```diff
@@ -1426,6 +1426,120 @@ def run_pass(self, state):
         return mutated
 
 
+@register_pass(mutates_CFG=False, analysis_only=False)
+class PropagateLiterals(FunctionPass):
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/core/untyped_passes.py:1431]: Right now it is specific to isinstance.

**Comment by stuartarchibald** [comment]
@guilhermeleobas As discussed OOB I've added some changes to this PR here:
https://github.com/stuartarchibald/numba/commits/pr_7244_continued
* https://github.com/stuartarchibald/numba/commit/8dae2fa78940a97ea71699ba20cc3582be0c8026 Adds in the experimental warning, a type guard and fixes an error message to use the unversioned var names.
* https://github.com/stuartarchibald/numba/commit/60ad609964af9640a4d250a8d61c346c0ef60763 Refactors the tests a bit and does another error message fix to use the unversioned var names.
* https://github.com/stuartarchibald/numba/commit/eb417f768eec937e94c1db305eacf930ff9c2fa1 is the largest patch and includes numerous tests. There are some issues with CPython scalar/NumPy scalar aliasing. Also some issues with the use of Numba types, like `types.Array` whereby removing the JIT decorator gives a different answer (`np.ndarray` instances are only instances of `types.Array` types under JIT compilation).
* https://github.com/stuartarchibald/numba/commit/dfa89450577785fb692e64ef974d47d12b34bce9 is a minor doc update.
* https://github.com/stuartarchibald/numba/commit/f2fd7a72f86b947b6de6917fa6c59345c3cc95ed removes DCE from this untyped pass as it's unsafe under aliasing.

If you'd be so kind as to take a look at the patches and if you think they are good then cherry-pick them to here, or if you have questions then please shout and I'll open a PR against your fork and we can discuss on there.

Left on my TODO for this PR is fixing up all the things @sklam noted RE updating the branch pruner to handle SSA.

**Comment by guilhermeleobas** [comment]
@stuartarchibald, your code seems good to me. I only have one comment regarding the `assert 0` statement at the end of the `test_branch_prune` test. This statement doesn't seem to be supported by Numba:

```python
numba/tests/test_builtins.py:1373: in test_branch_prune
    got = foo(x)
numba/core/dispatcher.py:494: in _compile_for_args
    error_rewrite(e, 'constant_inference')
numba/core/dispatcher.py:423: in error_rewrite
    raise e.with_traceback(None)
E   numba.core.errors.ConstantInferenceError: Failed in nopython mode pipeline (step: nopython rewrites)
E   Constant inference not possible for: call $88load_method.10(@py_format2, func=$88load_method.10, args=[Var(@py_format2, test_builtins.py:1369)], kws=(), vararg=None, target=None)
E   
E   File "numba/tests/test_builtins.py", line 1369:
E           def foo(x):
E               <source elided>
E               else:
E                   assert 0
E                   ^
E   
```

File: numba/core/analysis.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/analysis.py:504]: I think it's still needed as there could be branches pruned whilst in the SSA form and the `phi` nodes would need patching up as a result. This seems to trigger it:
```python
from numba import njit

@njit
def impl(pred=None, pred2=None):
    x = 1
    y = 10
    z = 100
    if pred2 is None:
        if pred is None:
            y = 9
            z = 12
        else:
            y = 102
            z = 13
    elif z < 22: # This dead, z is 100
        y = 91
    else:
        z = 7
        if pred is None:
            y = 7

    if isinstance(x, int):
        z = 987
    else:
        y = 52
    return x, y

print(impl(None, 1))
```

**Comment by stuartarchibald** [comment]
@guilhermeleobas I've added a couple of tests to make sure that the phi-node rewrite after branch pruning works, these are in https://github.com/stuartarchibald/numba/commit/eab51891a2b7348ba4096f5c1a5d813ecc62bdd3. Could please take a look and if you are happy with them put them onto this branch? Many thanks.

@sklam the above mentioned patch should hopefully address explicit testing of the phi node updates made in branch pruning.

**Comment by sklam** [comment]
> @sklam the above mentioned patch should hopefully address explicit testing of the phi node updates made in branch pruning.

New tests look good.

It's just the odd windows CI failure now.

**Comment by stuartarchibald** [comment]
Am really struggling to reproduce this failure locally. For reference the error is:
```
FAIL: test_diamond_2 (numba.tests.test_refop_pruning.TestRefOpPruning)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\1\s\numba\tests\test_refop_pruning.py", line 95, in test_diamond_2
    fanout=False, fanout_raise=False)
  File "D:\a\1\s\numba\tests\test_refop_pruning.py", line 49, in check
    self.assertEqual(stat, 0, msg=msg)
AssertionError: 7 != 0 : failed checking fanout_raise
```

**Comment by sklam** [comment]
RE: https://github.com/numba/numba/pull/7244#issuecomment-989762859

That failure does not consistently replicate on Azure. Let's just ignore it and see how it goes when everything is merged.

**Review by sklam — APPROVED** [review]

**Comment by stuartarchibald** [comment]
#7643 tracks the intermittent failure.
